### DISCUSSION TITLE
Architecture design record

### DISCUSSION THREAD

**Discussion by dhruvjain99** [wind-c/comqtt#29] [General]
@wind-c Is redis optional for using comqtt in cluster mode?

**Comment by dhruvjain99** [comment]
https://github.com/mochi-co/mqtt/discussions/174#discussioncomment-5276266
In reference to your comment, where do we store the subscription table? Is it stored on redis or on individual go application(comqtt) nodes?

**Reply by wind-c** [reply]
Subscription tables are stored in memory at each node and are synchronized and recovered through the raft protocol.

**Reply by dhruvjain99** [reply]
Then, is redis optional for using comqtt in cluster mode?

**Reply by wind-c** [reply]
redis is used to store the client's subscription history across nodes, as well as inflight and retain messages.The subscription routing table is used to route messages between nodes.For example, multiple clients on a node subscribe to the same filter, which is only used as one route.But from a client perspective, each subscription is stored in redis.
